File: src/main/java/com/ftn/sr192024/messenger/models/Message.java
package com.ftn.sr192024.messenger.models;

import com.fasterxml.jackson.annotation.JsonIgnore;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.hibernate.annotations.CreationTimestamp;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

@Entity
@Table(name = "messages")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
public class Message {
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @Column(columnDefinition = "TEXT")
    private String content;

    @CreationTimestamp
    @Column(name = "date_of_sending", nullable = false)
    private LocalDateTime dateOfSending;

    @ManyToOne
    @JoinColumn(name = "sender")
    private User sender;

    @ManyToOne
    @JoinColumn(name = "reply_to_message_id")
    private Message replyTo;

    @OneToMany(mappedBy = "message")
    private List<MessageReaction> reactions;

    @ManyToOne
    @JoinColumn(name = "forwarded_from_message_id")
    private Message forwardedFrom;

    @ManyToOne(optional = false)
    @JoinColumn(name = "chat_id")
    private Chat chat;

    @OneToMany(mappedBy = "message", cascade = CascadeType.ALL, orphanRemoval = true)
    @JsonIgnore
    private List<MessageReadStatus> readStatuses = new ArrayList<>();

    @Enumerated(EnumType.STRING)
    @Column(name = "message_type", nullable = false)
    private MessageType type = MessageType.USER;
}

File: chat-app/src/app/models/message.ts
import {User} from './user';
import {Chat} from './chat';
import {MessageReaction} from './reaction';

export type MessageStatus = 'SENT' | 'DELIVERED' | 'READ';

export interface Message {
  id: string;
  content: string;
  dateOfSending: string;
  sender: User;
  senderId: string;
  senderUsername: string;
  chat: Chat;
  replyTo: Message | null;
  forwardedFrom: Message | null;
  reactions: MessageReaction[];
  own?: boolean;
  status:MessageStatus
  deliveredAt?: string
  isSaved: boolean;
  type: 'USER' | 'SYSTEM'
}
